### ISSUE TITLE
Cache key for two equivalent requests differs when using indices boost

### ISSUE THREAD

**Issue by javanna** [elastic/elasticsearch#19986]
This is something I noticed when working on #19977.

Adding `assertEquals(deserializedRequest.cacheKey(), shardSearchTransportRequest.cacheKey());` to `ShardSearchTransportRequestTests#testSerialization` makes the test fail in some cases, seems due to the fact that indices boosts are stored in a map and the order of the items is different in the two requests that we are comparing. The cache key should be the same though, we should fix that.

**Comment by chengpohi** [comment]
@javanna I am trying to fix this bug, after investigation, Yes, this bug is caused by the order of `indexBoost:ObjectFloatHashMap` has different order when deserialised from `StreamInput`.

the caused `indexBoost` has different order even though the datas are same is `ObjectFloatHashMap` is using a `keyMixer` random number to calculate the `slot`.

```
  protected  int hashKey(KType key) {
    assert !((key) == null); // Handled as a special case (empty slot marker).
    return BitMixer.mix(key, this.keyMixer);
  }
  int slot = hashKey(existing) & mask;
```

when new a `ObjectFloatHashMap` it will create a new random `keyMixer`(`RandomizedHashOrderMixer`):

```
  protected final AtomicLong seedMixer;
  public RandomizedHashOrderMixer() {
    this(Containers.randomSeed64());
  }
  public RandomizedHashOrderMixer(long seed) {
    seedMixer = new AtomicLong(seed);
  }
  @Override
  public int newKeyMixer(int newContainerBufferSize) {
    return (int) BitMixer.mix64(seedMixer.incrementAndGet());
  }
```

So it's nearly impossible to keep `keyMixer` same, after `StreamIn`, I am trying to sort the `ObjectFloatHashMap` when `StreamOut`, Do you think we can do it? or Is there a better way to do it?

There is a PR for sorting `indexBoost` when `StreamOut`:

https://github.com/elastic/elasticsearch/pull/20081

**Comment by nik9000** [comment]
Fixed by #20081
